tests/disaster_recovery_3.py

```diff
@@ -0,0 +1,131 @@
+pnodes=4
+delay=args.d
+debug=args.v
+prod_count = 1 # per node prod count
+total_nodes=pnodes
+dumpErrorDetails=args.dump_error_details
+
+Utils.Debug=debug
+testSuccessful=False
+
+cluster=Cluster(unshared=args.unshared, keepRunning=args.leave_running, keepLogs=args.keep_logs)
+walletMgr=WalletMgr(True, keepRunning=args.leave_running, keepLogs=args.keep_logs)
+
+try:
+    TestHelper.printSystemInfo("BEGIN")
+
+    cluster.setWalletMgr(walletMgr)
+
+    Print(f'producing nodes: {pnodes}, delay between nodes launch: {delay} second{"s" if delay != 1 else ""}')
+
+    Print("Stand up cluster")
+    if cluster.launch(pnodes=pnodes, totalNodes=total_nodes, totalProducers=pnodes, delay=delay, loadSystemContract=False,
+                      activateIF=True, biosFinalizer=False) is False:
+        errorExit("Failed to stand up eos cluster.")
+
+    assert cluster.biosNode.getInfo(exitOnError=True)["head_block_producer"] != "eosio", "launch should have waited for production to change"
+    cluster.biosNode.kill(signal.SIGTERM)
+    cluster.waitOnClusterSync(blockAdvancing=5)
+
+    node0 = cluster.getNode(0) # A
+    node1 = cluster.getNode(1) # B
+    node2 = cluster.getNode(2) # C
+    node3 = cluster.getNode(3) # D
+
+    Print("Create snapshot (node 0)")
+    ret = node0.createSnapshot()
+    assert ret is not None, "Snapshot creation failed"
+    ret_head_block_num = ret["payload"]["head_block_num"]
+    Print(f"Snapshot head block number {ret_head_block_num}")
+
+    Print("Wait for snapshot node lib to advance")
+    assert node0.waitForBlock(ret_head_block_num+1, blockType=BlockType.lib), "Node0 did not advance to make snapshot block LIB"
+    assert node1.waitForLibToAdvance(), "Node1 did not advance LIB after snapshot of Node0"
+
+    assert node0.waitForLibToAdvance(), "Node0 did not advance LIB after snapshot"
+    currentLIB = node0.getIrreversibleBlockNum()
+    libBlock = node0.getBlock(currentLIB)
+    Print(f"Lib Block: {libBlock}")
+
+    Print("Shutdown all nodes")
+    for node in [node0, node1, node2, node3]:
+        node.kill(signal.SIGTERM)
+    for node in [node0, node1, node2, node3]:
+        assert not node.verifyAlive(), "Node did not shutdown"
+
+    Print("Remove reversible blocks and state, but not finalizers safety data")
+    for node in [node0, node1, node2, node3]:
+        node.removeReversibleBlks()
+        node.removeState()
+
+    Print("Restart nodes from snapshot")
+    for i in range(4):
+        isRelaunchSuccess = cluster.getNode(i).relaunch(chainArg=" -e --snapshot {}".format(node0.getLatestSnapshot()))
+        assert isRelaunchSuccess, f"node {i} relaunch from snapshot failed"
+
+    Print("Verify LIB does not advance on any node")
+    for node in [node0, node1, node2, node3]:
+        assert not node.waitForLibToAdvance(), "Node advanced LIB after relaunch when it should not"
+
+    Print("Shutdown all nodes to remove finalizer safety data")
```

Comment:
You don't remove finalizer safety data here.

Reply:
We do on line 112, that is why we are shutting down.